libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -0,0 +1,478 @@
+import Control.Monad.Trans.Class (MonadTrans (lift))
+import Control.Monad.Trans.RWS.Strict (RWST (..), ask, get, modify, runRWST)
+import Data.Default.Class (Default (def))
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe.Strict (StrictMaybe (SJust, SNothing))
+import qualified Data.Sequence.Strict as Seq
+import qualified Data.UMap as UM
+import Numeric.Natural
+import Test.Cardano.Ledger.Alonzo.Scripts (alwaysFails, alwaysSucceeds)
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import Test.Cardano.Ledger.Babbage.Serialisation.Generators ()
+import Test.Cardano.Ledger.Generic.Fields
+import Test.Cardano.Ledger.Generic.PrettyCore
+  ( dataHashSummary,
+    dataSummary,
+    keyHashSummary,
+    keyPairSummary,
+    scriptHashSummary,
+    scriptSummary,
+  )
+import Test.Cardano.Ledger.Generic.Proof hiding (lift)
+import Test.Cardano.Ledger.Generic.Updaters
+import Test.Tasty.QuickCheck
+  ( Gen,
+    Positive (..),
+    arbitrary,
+    choose,
+    chooseInt,
+    elements,
+    frequency,
+    generate,
+  )
+
+-- =================================================
+
+-- | Constants that determine how big a GenState is generated.
+data GenSize = GenSize
+  { numKeys :: Int,
+    numScripts :: Int,
+    numDatums :: Int,
+    numRewards :: Int,
+    numPools :: Int
+  }
+
+data GenEnv era = GenEnv
+  { geValidityInterval :: ValidityInterval,
+    gePParams :: Core.PParams era
+  }
+
+data GenState era = GenState
+  { gsKeys :: Map (KeyHash 'Witness (Crypto era)) (KeyPair 'Witness (Crypto era)),
+    gsScripts :: Map (ScriptHash (Crypto era)) (Core.Script era),
+    gsPlutusScripts :: Map (ScriptHash (Crypto era), Tag) (IsValid, Core.Script era),
+    gsDatums :: Map (DataHash (Crypto era)) (Data era),
+    gsDPState :: DPState (Crypto era)
+  }
+
+emptyGenState :: GenState era
+emptyGenState = GenState mempty mempty mempty mempty def
+
+instance Default GenSize where
+  def = GenSize {numKeys = 2, numScripts = 2, numDatums = 5, numRewards = 5, numPools = 1}
+
+-- ========================================================================
+-- Tools to get started
+
+initState :: Reflect era => Proof era -> GenSize -> GenRS era (GenState era)
+initState proof gs = do
+  replicateM_ (numKeys gs) genKeyHash
+  replicateM_ (numScripts gs) $ sequence_ [genScript proof i | i <- [Spend, Tag.Mint, Cert, Rewrd]]
+  replicateM_ (numDatums gs) genDatumWithHash
+  replicateM_ (numRewards gs) genRewards
+  replicateM_ (numPools gs) genPool
+  get
+
+startGen :: Reflect era => Proof era -> GenSize -> Gen (GenEnv era, GenState era)
+startGen proof gsize = do
+  env <- genGenEnv proof
+  (state, _, _) <- runRWST (initState proof gsize) env emptyGenState
+  pure (env, state)
+
+startGenRS :: Reflect era => Proof era -> GenSize -> GenRS era (GenEnv era, GenState era)
+startGenRS proof gsize = lift $ startGen proof gsize
+
+viewGenState :: Reflect era => Proof era -> IO ()
```

Comment:
I think it is good to annotate functions like that to inform the reader that they aren't suppose to be used in actual tests.
```suggestion
-- | Helper function for debugging the state generator inside ghci
viewGenState :: Reflect era => Proof era -> IO ()
```